src/hotspot/share/nmt/memLogRecorder.hpp

```diff
@@ -0,0 +1,212 @@
+
+public:
+  static void initialize();
+  static void finish();
+  static void replay();
+  static void logThreadName(const char* name);
+
+public:
+  void init();
+  void get_thread_name(char* buf);
+  bool done() {
+    return _done;
+  }
+  void logThreadName();
+};
+
+class NMT_MemoryLogRecorder : public NMT_LogRecorder {
+
+private:
+  static NMT_MemoryLogRecorder _recorder;
+
+private:
+    struct Entry {
+    long int time;
+    long int thread;
+    address ptr;
+    address old;
+    address stack[NMT_TrackingStackDepth];
+    long int requested;
+    long int actual;
+    long int mem_tag;
+  };
+
+public:
+  static NMT_MemoryLogRecorder* instance() {
+    return &_recorder;
+  };
+  static void initialize(long int count);
+  static bool initialized() {
+    return false;
+  }
+  static void print(Entry *e);
+  static void finish(void);
+  static void replay(const int pid);
+  static void record_free(void *ptr);
+  static void record_malloc(MemTag mem_tag, size_t requested, void* ptr, const NativeCallStack *stack, void* old = nullptr);
+  static void printActualSizesFor(const char* list);
+
+private:
+  static void _record(MemTag mem_tag, size_t requested, address ptr, address old, const NativeCallStack *stack);
+};
+
+class NMT_VirtualMemoryLogRecorder : public NMT_LogRecorder {
+
+private:
+  static NMT_VirtualMemoryLogRecorder _recorder;
+
+private:
+  struct Entry {
+    long int time;
+    long int thread;
+    address ptr;
+    address stack[NMT_TrackingStackDepth];
+    long int mem_tag;
+    long int mem_tag_split;
+    size_t size;
+    size_t size_split;
+    int type;
+  };
+
+public:
+  enum Type {
+    RESERVE,
+    RELEASE,
+    UNCOMMIT,
+    RESERVE_AND_COMMIT,
+    COMMIT,
+    SPLIT_RESERVED,
+    TAG
+  };
+
+public:
+  static NMT_VirtualMemoryLogRecorder* instance() {
+    return &_recorder;
+  };
+  static void initialize(long int count);
+  static void finish(void);
+  static void replay(const int pid);
+  static void record_virtual_memory_reserve(void* addr, size_t size, const NativeCallStack& stack, MemTag mem_tag = mtNone);
+  static void record_virtual_memory_release(address addr, size_t size);
+  static void record_virtual_memory_uncommit(address addr, size_t size);
+  static void record_virtual_memory_reserve_and_commit(void* addr, size_t size, const NativeCallStack& stack, MemTag mem_tag = mtNone);
+  static void record_virtual_memory_commit(void* addr, size_t size, const NativeCallStack& stack);
+  static void record_virtual_memory_split_reserved(void* addr, size_t size, size_t split, MemTag flag, MemTag mem_tag_split);
+  static void record_virtual_memory_tag(void* addr, size_t size, MemTag mem_tag);
+
+private:
+  static void _record(NMT_VirtualMemoryLogRecorder::Type type, MemTag mem_tag, MemTag mem_tag_split, size_t size, size_t size_split, address ptr, const NativeCallStack *stack);
+};
+
+#else // defined(LINUX) || defined(__APPLE__)
+
+class NMT_LogRecorder : public StackObj {
```

Comment:
What's the idea behind having two different subclasses for the log recorder? Like, why is it important that two different objects record the two sequences of events?

Reply:
Sorry, I am not clear on what you are asking. We have:

1. NMT_MemoryLogRecorder
2. NMT_VirtualMemoryLogRecorder

With more coming (ex: NMT_ArenasLogRecorder [NMT: add Arenas to NMTBenchmark](https://bugs.openjdk.org/browse/JDK-8353855) 

They share some common APIs, like init(), thread names, so it made sense to me to have both of them extend NMT_LogRecorder, which implements common functionality.